apps/gk_species.c

```diff
@@ -407,6 +409,97 @@ gk_species_apply_ic(gkyl_gyrokinetic_app *app, struct gk_species *gks, double t0
+
+    // Compute the polarization density.
+    struct gkyl_array *npol = mkarr(app->use_gpu, app->confBasis.num_basis, app->local_ext.volume);
+
+    struct gkyl_gyrokinetic_pol_density* npol_op = gkyl_gyrokinetic_pol_density_new(app->confBasis, app->grid, app->use_gpu);
+    gkyl_gyrokinetic_pol_density_advance(npol_op, &app->local, app->field->epsilon, app->field->phi_pol, npol);
+    gkyl_gyrokinetic_pol_density_release(npol_op);
+
+    // Calculate the guiding center density of this species: (npol - q_other*n^G_other)/q_self.
+    for (int i=0; i<app->num_species; ++i) {
+      struct gk_species *gks = &app->species[i];
+      gk_species_moment_calc(&gks->m0, gks->local, app->local, gks->f);
+      if (strcmp(gks->info.name, gks_self->info.name)) {
+        gkyl_array_accumulate(npol, -gks->info.charge, gks->m0.marr);
+      }
+    }
+    gkyl_array_scale(npol, 1./gks_self->info.charge);
+
+    // Scale the distribution function so it has this guiding center density.
+    struct gkyl_dg_bin_op_mem *div_mem = app->use_gpu? gkyl_dg_bin_op_mem_cu_dev_new(app->local.volume, app->confBasis.num_basis)
+                                                     : gkyl_dg_bin_op_mem_new(app->local.volume, app->confBasis.num_basis);
+    struct gkyl_array *den_mod = mkarr(app->use_gpu, app->confBasis.num_basis, app->local_ext.volume);
+
+    gkyl_dg_div_op_range(div_mem, app->confBasis,
+      0, den_mod, 0, npol, 0, gks_self->m0.marr, &app->local);
+    gkyl_dg_mul_conf_phase_op_range(&app->confBasis, &app->basis, gks_self->f,
+      den_mod, gks_self->f, &app->local_ext, &gks_self->local_ext);
+
+          
+    const char *fmt = "%s-%s.gkyl";
+    int sz = gkyl_calc_strlen(fmt, app->name, "den_mod"); 
+    char fileNm[sz+1];
+    int rank;
+    gkyl_comm_get_rank(app->comm, &rank);
+
+
+    struct gkyl_array *den_mod_global = mkarr(app->use_gpu, app->confBasis.num_basis, app->global_ext.volume);
+    struct gkyl_array *npol_global = mkarr(app->use_gpu, app->confBasis.num_basis, app->global_ext.volume);
+    struct gkyl_array *m0_global = mkarr(app->use_gpu, app->confBasis.num_basis, app->global_ext.volume);
+
+    gkyl_comm_array_allgather(app->comm, &app->local, &app->global, npol, npol_global);
+    gkyl_comm_array_allgather(app->comm, &app->local, &app->global, den_mod, den_mod_global);
+    gkyl_comm_array_allgather(app->comm, &app->local, &app->global, gks_self->m0.marr, m0_global);
+    struct gkyl_array *den_mod_ho, *npol_ho, *m0_ho;
+    if (app->use_gpu) {
+      den_mod_ho = mkarr(false, app->confBasis.num_basis, app->global_ext.volume);
+      npol_ho = mkarr(false, app->confBasis.num_basis, app->global_ext.volume);
+      m0_ho = mkarr(false, app->confBasis.num_basis, app->global_ext.volume);
+      gkyl_array_copy(den_mod_ho, den_mod_global);
+      gkyl_array_copy(npol_ho, npol_global);
+      gkyl_array_copy(m0_ho, m0_global);
+    } else {
+      den_mod_ho = den_mod;
+      npol_ho = npol;
+      m0_ho = gks_self->m0.marr;
+    }
```

Comment:
I don't recognize all this comm and global stuff. If you are doing it for I/O, please remove it, see comment below.